### ISSUE TITLE
Non-nullable promises shouldn't be allowable in conditions

### ISSUE THREAD

**Issue by travigd** [microsoft/TypeScript#34717]
## Search Terms
Promise, conditions, if, truthiness, coercsion, known-true

## Related
#9041, #33178

## Suggestion
Non-nullable promises should not be implicitly converted to truthy inside of conditions.

## Use Cases

This helps prevent common bugs where promises aren't `await`-ed (see below).

## Examples

```ts
async isAllowedToDoTheSuperSecretThing(user: User) {
    /* ... */
}

async doTheSuperSecretThing(ctx: RequestContext) {
  if (isAllowedToDoTheSuperSecretThing(ctx.user)) {
    /* do the super secret thing */
  }
}
```

should result in an error since `isAllowedToDoTheSuperSecretThing` returns a non-nullable promise and is never falsey.

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

This definitely falls into the camp of goal one:
**Statically identify constructs that are likely to be errors.**

This would be breaking but so is #33178.

**Comment by travigd** [comment]
@RyanCavanaugh This was tagged "Awaiting More Feedback." Does that have any specific meaning or does it just generically mean "this is on the back-burner"?

**Comment by travigd** [comment]
This is available as a plugin for ESLint: https://github.com/typescript-eslint/typescript-eslint/pull/612.

The downside is that it requires type checking in ESLint which is pretty slow since it generally requires the whole codebase to be checked (whereas when just developing with the language server running, it's pretty fast).

**Comment by andrewbranch** [comment]
Implemented in #39175
